lib/aot/compiler.cpp

```diff
@@ -2905,6 +3171,148 @@ class FunctionCompiler {
+    auto *OkBB = llvm::BasicBlock::Create(LLContext, "address_align_ok", F);
+    auto *IsAddressAligned =
+        createLikely(Builder, Builder.CreateICmpEQ(Value, Builder.getInt64(0)));
+    Builder.CreateCondBr(IsAddressAligned, OkBB,
+                         getTrapBB(ErrCode::UnalignedAtomicAccess));
+
+    Builder.SetInsertPoint(OkBB);
+  }
+
+  void compileMemoryFence() {
+    Builder.CreateFence(llvm::AtomicOrdering::SequentiallyConsistent);
+  }
+  void compileAtomicNotify() {
+    Builder.CreateCall(Context.getIntrinsic(
+                           Builder,
+                           AST::Module::Intrinsics::kMemoryAtomicNotify,
+                           llvm::FunctionType::get(Context.VoidTy, {}, false)),
+                       {});
+  }
+  void compileAtomicWait(unsigned MemIdx, unsigned BitWidth) {
+    auto *Offset = Builder.CreateSExtOrTrunc(stackPop(), Context.Int32Ty);
+    auto *ExpectedValue =
+        Builder.CreateSExtOrTrunc(stackPop(), Context.Int64Ty);
+    auto *Timeout = Builder.CreateSExtOrTrunc(stackPop(), Context.Int64Ty);
+
+    stackPush(Builder.CreateCall(
+        Context.getIntrinsic(
+            Builder, AST::Module::Intrinsics::kMemoryAtomicWait,
+            llvm::FunctionType::get(Context.Int32Ty,
+                                    {Context.Int32Ty, Context.Int32Ty,
+                                     Context.Int64Ty, Context.Int64Ty,
+                                     Context.Int32Ty},
+                                    false)),
+        {Builder.getInt32(MemIdx), Offset, ExpectedValue, Timeout,
+         Builder.getInt32(BitWidth)}));
+  }
+  void compileAtomicLoad(unsigned MemoryIndex, unsigned Alignment,
+                         llvm::IntegerType *IntType,
+                         llvm::IntegerType *TargetType, bool Signed = false) {
+
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
+    compileAtomicCheckOffsetAlignment(Offset, TargetType);
+    auto *VPtr = Builder.CreateInBoundsGEP(
+        Context.Int8Ty, Context.getMemory(Builder, ExecCtx, MemoryIndex),
+        Offset);
+
+    auto *Ptr = Builder.CreateBitCast(VPtr, TargetType->getPointerTo());
+    auto *Load = Builder.CreateLoad(TargetType, Ptr, OptNone);
+    Load->setAlignment(Align(UINT64_C(1) << Alignment));
+    Load->setAtomic(llvm::AtomicOrdering::SequentiallyConsistent);
+
+    if (Signed) {
+      Stack.back() = Builder.CreateSExt(Load, IntType);
+    } else {
+      Stack.back() = Builder.CreateZExt(Load, IntType);
+    }
+  }
+  void compileAtomicStore(unsigned MemoryIndex, unsigned Alignment,
+                          llvm::IntegerType *, llvm::IntegerType *TargetType,
+                          bool Signed = false) {
+
+    auto *V = stackPop();
+
+    if (Signed) {
+      V = Builder.CreateSExtOrTrunc(V, TargetType);
+    } else {
+      V = Builder.CreateZExtOrTrunc(V, TargetType);
+    }
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
```

Comment:
Final offset should add offset value in `instruction`.
offset = 'stack pop value' + 'instruction offset value'